### ISSUE TITLE
Suspicious infinite loop symptoms when `getDefaultDialect()` not defined on fallback interface.

### ISSUE THREAD

**Issue by stellaraccident** [google/iree#8186]
```
#0  0x0000000002575700 in mlir::detail::OpAsmOpInterfaceInterfaceTraits::FallbackModel<mlir::iree_compiler::IREE::VM::VMDialect::VMOpAsmInterface>::getDefaultDialect() ()                                           
#1  0x0000000003f4a27a in (anonymous namespace)::OperationPrinter::printRegion(mlir::Region&, bool, bool, bool) ()                                                                                                   
#2  0x0000000003f4a030 in (anonymous namespace)::OperationPrinter::printGenericOp(mlir::Operation*, bool) ()                                                                                                         
#3  0x0000000003f45537 in (anonymous namespace)::OperationPrinter::print(mlir::Operation*) ()                                                                                                                        
#4  0x0000000003f43cad in mlir::Operation::print(llvm::raw_ostream&, mlir::AsmState&, mlir::OpPrintingFlags const&) ()                                                                                               
#5  0x0000000003f43955 in mlir::Operation::print(llvm::raw_ostream&, mlir::OpPrintingFlags const&) ()                                                                                                                
#6  0x0000000003ef4b9b in mlir::Diagnostic::appendOp(mlir::Operation&, mlir::OpPrintingFlags const&) ()                                                                                                              
#7  0x0000000003ec78ca in mlir::Operation::emitError(llvm::Twine const&) ()                                                                                                                                          
#8  0x0000000003eb823a in mlir::detail::verifySymbolTable(mlir::Operation*) ()                                                                                                                                       
#9  0x0000000003f12156 in mlir::Op<mlir::ModuleOp, mlir::OpTrait::OneRegion, mlir::OpTrait::ZeroResult, mlir::OpTrait::ZeroSuccessor, mlir::OpTrait::ZeroOperands, mlir::OpTrait::AffineScope, mlir::OpTrait::IsIsola
tedFromAbove, mlir::OpTrait::NoRegionArguments, mlir::OpTrait::SymbolTable, mlir::SymbolOpInterface::Trait, mlir::OpAsmOpInterface::Trait, mlir::OpTrait::NoTerminator, mlir::OpTrait::SingleBlock, mlir::RegionKindI
nterface::Trait, mlir::OpTrait::HasOnlyGraphRegion>::verifyInvariants(mlir::Operation*) ()                                                                                                                           
#10 0x0000000003eb2cce in (anonymous namespace)::OperationVerifier::verifyOperation(mlir::Operation&, llvm::SmallVectorImpl<mlir::Operation*>&) ()                                                                   
#11 0x0000000003eb24f5 in (anonymous namespace)::OperationVerifier::verifyOpAndDominance(mlir::Operation&) ()                                                                                                        
#12 0x0000000003eb2494 in mlir::verify(mlir::Operation*) ()                                                                                                                                                          
#13 0x0000000003d42965 in mlir::parseSourceFile(llvm::SourceMgr const&, mlir::Block*, mlir::MLIRContext*, mlir::LocationAttr*, mlir::AsmParserState*) () 
```

To repro once https://github.com/google/iree/pull/8174 lands, from `VMDialect.cpp`, comment out `static StringRef getDefaultDialect() { return ""; }` from `VMOpAsmInterface`.

**Comment by stellaraccident** [comment]
Fails in multiple trivial lit tests. For example: `iree/compiler/Dialect/VM/IR/test/list_ops.mlir`

**Comment by joker-eph** [comment]
The implementation of the method at the top of your stack is:
```
template<typename ConcreteOp>
::llvm::StringRef detail::OpAsmOpInterfaceInterfaceTraits::FallbackModel<ConcreteOp>::getDefaultDialect() {
  return ConcreteOp::getDefaultDialect();
}
```

So it really forwards to `VMDialect::VMOpAsmInterface:: getDefaultDialect`, I'll be interested to know what this resolves to when the code is commented out! (that is: why it even builds!)

**Comment by stellaraccident** [comment]
Yes, I was surprised it builds: I did get a build error on the missing block argument name instance method originally (and then remembered what you said and added it). But I didn't see an error for `getDefaultDialect` and only went to look at the interface definition after the debugger told me it was hung there.

**Comment by joker-eph** [comment]
This is an issue with `StaticInterfaceMethod` and `FallbackModel`.

Here is the implementation of the "trampoline" in the interface:
```
template<typename ConcreteOp>
::llvm::StringRef detail::OpAsmOpInterfaceInterfaceTraits::FallbackModel<ConcreteOp>::getDefaultDialect() {
  return ConcreteOp::getDefaultDialect();
}
```

Here `ConcreteOp` is the fallback interface you implement, so it properly forward to it.
However, the fallback interface all inherit from this FallbackModel itself, so if it does not redefine the `getDefaultDialect` it inherit this implementation... which calls itself!!

**Comment by ScottTodd** [comment]
This is still an issue, but low priority since the workaround is just to keep one line of code 🤔 

> To repro once #8174 lands, from `VMDialect.cpp`, comment out `static StringRef getDefaultDialect() { return ""; }` from `VMOpAsmInterface`.

That line still exists, and `iree/compiler/Dialect/VM/IR/test/list_ops.mlir` still fails if I comment it out.
